Mypy incorrectly warns when passing literals as kwarg keywords

**Bug Report**

```
from typing import Any, Dict, Literal
kw: Dict[Literal["a", "b"], Any]
def func(a, b):
    pass
func(**kw)
```
Leads to
```
main.py:8: error: Keywords must be strings
```

https://mypy-play.net/?mypy=latest&python=3.9&gist=582cf5e8c0a6e6a5881d391da64e1455

**Expected Behavior**
Expect this to be allowable - as these literals count as strings

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -6133,8 +6133,17 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
 
     def is_valid_keyword_var_arg(self, typ: Type) -> bool:
         """Is a type valid as a **kwargs argument?"""
+        typ = get_proper_type(typ)
         return (
-            is_subtype(
+            (
+                # This is a little ad hoc, ideally we would have a map_instance_to_supertype
+                # that worked for protocols
+                isinstance(typ, Instance)
+                and typ.type.fullname == "builtins.dict"
+                and is_subtype(typ.args[0], self.named_type("builtins.str"))
+            )
+            or isinstance(typ, ParamSpecType)
+            or is_subtype(
                 typ,
                 self.chk.named_generic_type(
                     "_typeshed.SupportsKeysAndGetItem",
@@ -6147,7 +6156,6 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
                     "_typeshed.SupportsKeysAndGetItem", [UninhabitedType(), UninhabitedType()]
                 ),
             )
-            or isinstance(typ, ParamSpecType)
         )
 
     def not_ready_callback(self, name: str, context: Context) -> None:

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -1401,7 +1401,7 @@ class MessageBuilder:
     def invalid_keyword_var_arg(self, typ: Type, is_mapping: bool, context: Context) -> None:
         typ = get_proper_type(typ)
         if isinstance(typ, Instance) and is_mapping:
-            self.fail("Keywords must be strings", context)
+            self.fail("Argument after ** must have string keys", context, code=codes.ARG_TYPE)
         else:
             self.fail(
                 f"Argument after ** must be a mapping, not {format_type(typ, self.options)}",

--- a/test-data/unit/check-generic-subtyping.test
+++ b/test-data/unit/check-generic-subtyping.test
@@ -1014,7 +1014,7 @@ func_with_kwargs(**x1)
 [out]
 main:12: note: Revealed type is "typing.Iterator[builtins.int]"
 main:13: note: Revealed type is "builtins.dict[builtins.int, builtins.str]"
-main:14: error: Keywords must be strings
+main:14: error: Argument after ** must have string keys
 main:14: error: Argument 1 to "func_with_kwargs" has incompatible type "**X1[str, int]"; expected "int"
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-medium.pyi]

--- a/test-data/unit/check-kwargs.test
+++ b/test-data/unit/check-kwargs.test
@@ -334,7 +334,7 @@ def f(**kwargs: 'A') -> None: pass
 b: Mapping
 d: Mapping[A, A]
 m: Mapping[str, A]
-f(**d)         # E: Keywords must be strings
+f(**d)         # E: Argument after ** must have string keys
 f(**m)
 f(**b)
 class A: pass
@@ -354,7 +354,7 @@ from typing import Dict, Any, Optional
 class A: pass
 def f(**kwargs: 'A') -> None: pass
 d = {} # type: Dict[A, A]
-f(**d)         # E: Keywords must be strings
+f(**d)         # E: Argument after ** must have string keys
 f(**A())       # E: Argument after ** must be a mapping, not "A"
 kwargs: Optional[Any]
 f(**kwargs)    # E: Argument after ** must be a mapping, not "Any | None"
@@ -449,7 +449,7 @@ f(b) # E: Argument 1 to "f" has incompatible type "dict[str, str]"; expected "in
 f(**b) # E: Argument 1 to "f" has incompatible type "**dict[str, str]"; expected "int"
 
 c = {0: 0}
-f(**c) # E: Keywords must be strings
+f(**c) # E: Argument after ** must have string keys
 [builtins fixtures/dict.pyi]
 
 [case testCallStar2WithStar]
@@ -566,4 +566,14 @@ main:37: error: Argument 1 to "foo" has incompatible type "**B[str, str]"; expec
 main:38: error: Argument after ** must be a mapping, not "C[str, float]"
 main:39: error: Argument after ** must be a mapping, not "D"
 main:41: error: Argument 1 to "foo" has incompatible type "**dict[str, str]"; expected "float"
+[builtins fixtures/dict.pyi]
+
+[case testLiteralKwargs]
+from typing import Any, Literal
+kw: dict[Literal["a", "b"], Any]
+def func(a, b): ...
+func(**kw)
+
+badkw: dict[Literal["one", 1], Any]
+func(**badkw)  # E: Argument after ** must have string keys
 [builtins fixtures/dict.pyi]

